### PR TITLE
feat(generativeai): Add sample for Prompt Optimiser

### PR DESCRIPTION
**Pull request by Sita04**
Add sample for VertexAI Prompt Optimizer.

More details @ [b/366499370](http://b/366499370)

## Checklist
- [X] I have followed [Sample Guidelines from AUTHORING_GUIDE.MD](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/AUTHORING_GUIDE.md)
- [ ] README is updated to include [all relevant information](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/AUTHORING_GUIDE.md#readme-file)
- [ ] **Tests** pass:   `nox -s py-3.9` (see [Test Environment Setup](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/AUTHORING_GUIDE.md#test-environment-setup))
- [X] **Lint** pass:   `nox -s lint` (see [Test Environment Setup](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/AUTHORING_GUIDE.md#test-environment-setup))
- [ ] These samples need a new **API enabled** in testing projects to pass (let us know which ones)
- [ ] These samples need a new/updated **env vars** in testing projects set to pass (let us know which ones)
- [ ] This sample adds a new sample directory, and I updated the [CODEOWNERS file](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/.github/CODEOWNERS) with the codeowners for this sample
- [ ] This sample adds a new **Product API**, and I updated the [Blunderbuss issue/PR auto-assigner](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/.github/blunderbuss.yml) with the codeowners for this sample
- [X] Please **merge** this PR for me once it is approved

### CODE REVIEW

File: generative_ai/prompts/prompt_optimizer.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+
+
+#  [START generativeaionvertexai_prompt_optimizer]
```
Comment (rsamborski) [inline generative_ai/prompts/prompt_optimizer.py:18]: fix: According to latest recommendation (go/runnable-snippets) the opening tag should start inside the function to make the sample executable as-is (without a need to call the function)

File: generative_ai/prompts/prompt_optimizer.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        project: Google Cloud Project ID.
+        location: Location where you want to run the Vertex AI prompt optimizer.
+        staging_bucket: Specify the Google Cloud Storage bucket to store outputs and metadata. For example, gs://bucket-name
+        configuration_path: URI of the configuration file in your Google Cloud Storage bucket. For example, gs://bucket-name/configuration.json.
+    Returns:
+        custom_job.resource_name: Returns the resource name of the job created of type: projects/project-id/locations/location/customJobs/job-id
+    """
+    from google.cloud import aiplatform
+
+    aiplatform.init(project=project, location=location, staging_bucket=staging_bucket)
```
Comment (rsamborski) [inline generative_ai/prompts/prompt_optimizer.py:36]: fix: Please add a section with variables that should be updated and commented our by the user ([example](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/9aa31c1e1defc09f1392fe09db4332221fb68bbe/generative_ai/function_calling/chat_function_calling_basic.py#L27))

File: generative_ai/prompts/prompt_optimizer.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+            },
+        }
+    ]
+
+    custom_job = aiplatform.CustomJob(
+        display_name="Prompt Optimizer example",
+        worker_pool_specs=worker_pool_specs,
+    )
+    custom_job.submit()
+    return custom_job.resource_name
```
Comment (rsamborski) [inline generative_ai/prompts/prompt_optimizer.py:56]: fix: Please make sure to print out the result of running the function and keep return after the region tag is closed

File: generative_ai/prompts/test_prompt_optimizer.py [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+        "sample_prompt_template.txt",
+        "sample_prompts.jsonl",
+        "sample_system_instruction.txt",
+    ]
+    # cleanup existing stale resources
+    _clean_resources(STAGING_BUCKET_NAME)
+    # create bucket
+    bucket = STORAGE_CLIENT.bucket(STAGING_BUCKET_NAME)
+    bucket.storage_class = "STANDARD"
+    new_bucket = STORAGE_CLIENT.create_bucket(bucket, location="us")
```
Comment (rsamborski) [inline generative_ai/prompts/test_prompt_optimizer.py:89]: suggestion: Please consider creating the bucket only if it doesn't exist. With current approach we risk that someone will grab our bucket name (name must be globally unique) and the test will start failling.

**Review by rsamborski — CHANGES_REQUESTED** [review]

File: generative_ai/prompts/test_prompt_optimizer.py (hunk 4 above)
Comment (Sita04) [reply generative_ai/prompts/test_prompt_optimizer.py:89]: Great catch!

I was weighing against having the bucket as a permanent resource vs creating it just for the test. Agree that there could potentially be a name clash even if its implemented dynamically.

Looks like the better approach is to upload static config files into a bucket in all the CI projects? If so, kindly help with provisioning.

File: generative_ai/prompts/prompt_optimizer.py (hunk 1 above)
Comment (Sita04) [reply generative_ai/prompts/prompt_optimizer.py]: done!

**Comment by msampathkumar** [comment]
I added `DONOTMERGE` to merged the PR at the advised time.

File: generative_ai/prompts/test_prompt_optimizer.py (hunk 4 above)
Comment (rsamborski) [reply generative_ai/prompts/test_prompt_optimizer.py:93]: SG, please sync with Sampath on this. Nevertheless I would still keep the files in the source code as a reference for people who download the sample/clone the repo.

**Review by rsamborski — COMMENTED** [review]
Unfortunately the tests are failing with following error:
`RuntimeError: location us-west4 is provided, but different from the resource location us-central1`

**Review by rsamborski — APPROVED** [review]

**Comment by msampathkumar** [comment]
New error: `RuntimeError: location us-west1 is provided, but different from the resource location us-central1 `

Suggestion: To use only `us-central1` location

**Review by msampathkumar — APPROVED** [review]
LGTM

**Review by m-strzelczyk — APPROVED** [review]

**Pull request merged**